Convert DateTime to PHP value using Doctrine converter

k0ka opened the pull request:
Convert DateTime to PHP value using Doctrine converter. It allows using Timestampable with other date objects like Carbon.

Fixes https://github.com/doctrine-extensions/DoctrineExtensions/issues/1946

I guess the same should be done to `SoftDeleteableAdapter`. We will have to add the `nesbot/carbon` dev dependency to fully test it.

phansys commented on src/Timestampable/Mapping/Event/Adapter/ORM.php:
```diff
@@ -20,6 +22,20 @@ final class ORM extends BaseAdapterORM implements TimestampableAdapter
+     * Generates current timestamp for the specified mapping
+     *
+     * @param object $mapping
+     * @return mixed
```
```suggestion
     *
     * @return \DateTimeInterface|int
```

phansys commented on src/Timestampable/Mapping/Event/Adapter/ORM.php:
```diff
@@ -20,6 +22,20 @@ final class ORM extends BaseAdapterORM implements TimestampableAdapter
+    /**
+     * Generates current timestamp for the specified mapping
+     *
+     * @param object $mapping
```
```suggestion
     * @param array<string, mixed> $mapping
```

phansys commented on src/Timestampable/Mapping/Event/Adapter/ORM.php:
```diff
@@ -20,6 +22,20 @@ final class ORM extends BaseAdapterORM implements TimestampableAdapter
+     * @param object $mapping
+     * @return mixed
+     */
+    public function getRawDateValue($mapping)
```
```suggestion
    private function getRawDateValue(array $mapping)
```

phansys commented on src/Timestampable/Mapping/Event/Adapter/ORM.php:
```diff
@@ -30,4 +46,5 @@ public function getDateValue($meta, $field)
         return \DateTime::createFromFormat('U.u', number_format(microtime(true), 6, '.', ''))
             ->setTimeZone(new \DateTimeZone(date_default_timezone_get()));
     }
+
```
```suggestion
```

k0ka commented:
Hi. 

1. Added Carbon support to SoftDeletable
2. Added Carbon as a dev dependency to test
3. Added tests for Timestampable/SoftDeletable
4. Added ChangeLog record

Please let me know If something needs changing

phansys commented on CHANGELOG.md:
```diff
@@ -23,6 +23,8 @@ a release.
 - SoftDeleteable: Support to use annotations as attributes on PHP >= 8.0.
 - Blameable: Support to use annotations as attributes on PHP >= 8.0.
 - IpTraceable: Support to use annotations as attributes on PHP >= 8.0.
+- SoftDeleteable: Support for Carbon type attributes
```
I'm not a Carbon user. Are these changes only compatible with the features provided by this library or with any other custom implementation?

k0ka replied:
Any custom implementation. But I guess Carbon is the most frequently used.

phansys replied:
Then, I think we should be clear about that.
Consider something like this:
>SoftDeleteable: Support for type attributes in custom `\DateTimeInterface` implementations (like [Carbon](https://carbon.nesbot.com)).

k0ka replied:
I added a clarification. 
```Support for custom column types (like Carbon).```
These columns might be of any type, like someone could write a simple converter to store it as a string.

phansys commented on tests/Gedmo/Timestampable/CarbonTest.php:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    /**
+     * @test
```
Please, use "test" as method prefix instead of this annotation.

k0ka replied:
done

franmomu commented on tests/Gedmo/SoftDeleteable/CarbonTest.php:
```diff
@@ -0,0 +1,107 @@
+use Gedmo\Tests\SoftDeleteable\Fixture\Entity\Comment;
+use Gedmo\Tests\Tool\BaseTestCaseORM;
+
+class CarbonTest extends BaseTestCaseORM
```
```suggestion
final class CarbonTest extends BaseTestCaseORM
```

franmomu commented on tests/Gedmo/Timestampable/CarbonTest.php:
```diff
@@ -0,0 +1,160 @@
+use Gedmo\Tests\Tool\BaseTestCaseORM;
+use Gedmo\Timestampable\TimestampableListener;
+
+class CarbonTest extends BaseTestCaseORM
```
```suggestion
final class CarbonTest extends BaseTestCaseORM
```

franmomu commented on tests/Gedmo/Timestampable/CarbonTest.php:
```diff
@@ -0,0 +1,160 @@
+        static::assertNotSame($sport->getAuthorChanged(), $sa, 'Author must have changed after update');
+    }
+
+    protected function getUsedEntityFixtures()
```
```suggestion
    protected function getUsedEntityFixtures(): array
```

franmomu commented on tests/Gedmo/Timestampable/CarbonTest.php:
```diff
@@ -0,0 +1,160 @@
+        DoctrineType::overrideType(Types::DATETIME_MUTABLE, \Doctrine\DBAL\Types\DateTimeType::class);
+    }
+
+    public function testShouldHandleStandardBehavior()
```
```suggestion
    public function testShouldHandleStandardBehavior(): void
```

franmomu commented on tests/Gedmo/SoftDeleteable/CarbonTest.php:
```diff
@@ -0,0 +1,107 @@
+        static::assertInstanceOf(Carbon::class, $comment->getDeletedAt());
+    }
+
+    protected function getUsedEntityFixtures()
```
```suggestion
    protected function getUsedEntityFixtures(): array
```

phansys commented:
@olml89, could you please confirm if #2273 can be closed in favor of this PR?

phansys commented on tests/Gedmo/SoftDeleteable/CarbonTest.php:
```diff
@@ -0,0 +1,107 @@
+use Gedmo\Tests\SoftDeleteable\Fixture\Entity\Comment;
+use Gedmo\Tests\Tool\BaseTestCaseORM;
+
+final class CarbonTest extends BaseTestCaseORM
```
IMO, these classes should not include "Carbon" in its names, but something like "CustomDateTime".

k0ka replied:
Well, I can rename these files.

However they not only test custom field types but also check integration with Carbon converter.

`CustomDateTime` might be confusing too as `DateTime` is one of Doctrine types for physical time. Others are `Time` and `Date` (and all are postfixed by `_immutable` or `_mutable`). And custom converter should not be always of any time type. We can define a new custom type `my_timestamp` that will be stored in the database as a string or blob. And it will be handled by the same code.

So `CustomFieldType` would make more sense, but there is only one simple test there. Is it reasonable to discuss its name?

phansys replied:
I think these are good concerns. The goal of my suggestion is to keep these files agnostic as possible in order to allow more tests in the future for custom implementations other than Carbon.
IMO, if the integration with the Carbon converter can be made separately, I think it could be a good decision, as
it is always better to keep the responsibilities separated.

k0ka replied:
Uhm, I don't get it. What should I do?

phansys replied:
>However they not only test custom field types but also check integration with Carbon converter.

>Uhm, I don't get it. What should I do?

From my perspective, the test should be organized in this way:
1. Any test related to the custom types should cover the responsibilities related to a custom type implementation (Carbon or others).
2. If we have a exclusive test covering the integration with the Carbon converter, I think this test should be provided in a separate class, as it is only related to the Carbon package.

Please, let me know if you need more context about my suggestion.

k0ka replied:
There are only two tests (one for each extension). They are just copies of tests which check the main behavior with added field type assertions. 

In order to test this feature we need a Doctrine Field Converter. We may write a fake one by our own but we will have to maintain it and track when Doctrine change any requirements. Or we may use quite popular and maintained converter, the one provided by the Carbon team.

I've chosen the latter option.

k0ka commented:
I did a rebase and fixed errors. Should I do anything else?

phansys commented:
Thanks @k0ka!
